File: src/network/models/http_request.py
from __future__ import annotations
import base64
from dataclasses import dataclass
import json

from network.models.flow_request import FlowRequest
from shared.model import Model


@dataclass(kw_only=True)
class HttpRequest(FlowRequest):
    method: str
    host: str
    path: str
    http_version: str
    headers: dict[str, list[str]]
    body: str

    meta = {
        "relationship_keys": [
            "request",
            "response",
        ],
        "json_columns": [],
        "do_not_save_keys": [],
    }

    @classmethod
    def from_json(cls, json_raw: bytes) -> HttpRequest:
        values = json.loads(json_raw)
        return HttpRequest(
            method=values['method'],
            path=values['path'],
            host=values['host'],
            http_version=values['http_version'],
            headers=values['headers'],
            body=values['body'],
        )

    def to_json(self) -> str:
        return json.dumps(self.values_for_db())

    def __str__(self):
        out = f"{self.method} {self.path} HTTP/{self.http_version}\n"
        for key, value in self.headers.items():
            v = ",".join(value)
            out += f"{key}: {v}\n"

        out += "\r\n"
        if len(self.body) > 0:
            out += self.body

        return out

File: lib/network/models/http_request.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:trayce/common/utils.dart';
import 'package:trayce/network/models/flow_request.dart';

import '../../agent/gen/api.pb.dart' as pb;

class HttpRequest extends FlowRequest {
  final String method;
  final String host;
  final String path;
  final String httpVersion;
  final Map<String, List<String>> headers;
  final String body;

  const HttpRequest({
    required this.method,
    required this.host,
    required this.path,
    required this.httpVersion,
    required this.headers,
    required this.body,
  }) : super();

  /// Creates an HttpRequest from raw JSON bytes
  static HttpRequest fromJson(Uint8List bytes) {
    final values = json.decode(utf8.decode(bytes)) as Map<String, dynamic>;
    return HttpRequest(
      method: values['method'] as String,
      path: values['path'] as String,
      host: values['host'] as String,
      httpVersion: values['http_version'] as String,
      headers: (values['headers'] as Map<String, dynamic>).map(
        (key, value) => MapEntry(key, (value as List).cast<String>()),
      ),
      body: values['body'] as String,
    );
  }

  /// Creates an HttpRequest from a protobuf HTTPRequest
  factory HttpRequest.fromProto(pb.HTTPRequest req) {
    // Convert headers from protobuf repeated fields to Map<String, List<String>>
    final headers = <String, List<String>>{};
    for (var entry in req.headers.entries) {
      headers[entry.key] = entry.value.values.map((v) => v.toString()).toList();
    }

    // Convert body from bytes to UTF-8 string if present
    final body = req.payload.isEmpty ? '' : utf8.decode(req.payload);

    return HttpRequest(
      method: req.method,
      path: req.path,
      host: req.host,
      httpVersion: req.httpVersion,
      headers: headers,
      body: body,
    );
  }

  @override
  Uint8List toJson() {
    final map = {
      'method': method,
      'path': path,
      'host': host,
      'http_version': httpVersion,
      'headers': headers,
      'body': body,
    };
    return utf8.encode(json.encode(map));
  }

  @override
  String toString() {
    var out = '$method $path HTTP/$httpVersion\n';

    out += formatSortedHeaders(headers);

    out += '\r\n';
    if (body.isNotEmpty) {
      out += body;
    }

    return out;
  }

  @override
  String operationCol() {
    return "$method $path";
  }
}
